server.go

```diff
@@ -187,3 +224,196 @@ func copyWithHeader(r io.Reader, w io.Writer, header proto.Header) error {
+		sess.timeout = time.AfterFunc(options.SessionTimeout, func() {
+			// Delete immediately in case it takes a while to clean up otherwise new
+			// connections with this ID will try to connect to the closing screen.
+			sessions.Delete(id)
+			// Close() will send a SIGTERM allowing screen to clean up.  If we kill it
+			// abruptly the socket will be left behind which is probably harmless but
+			// it causes screen -list to show a bunch of dead sessions (of course the
+			// user would not see this since we have a custom socket directory) and it
+			// seems ideal to let it clean up anyway.
+			process.Close()
+			select {
+			case <-sess.ctx.Done():
+				return
+			case <-time.After(5 * time.Second):
+				cancel() // Force screen to exit.
+			}
+		})
+
+		// Remove the session if screen exits.
+		go func() {
+			err := process.Wait()
+			// Screen exits with a one when it gets a SIGTERM.
+			if exitErr, ok := err.(ExitError); ok && exitErr.ExitCode() != 1 {
+				flog.Error("session %s exited with error %w", id, err)
+			}
+			cancel()
+			sess.timeout.Stop()
+			sessions.Delete(id)
+		}()
+
+		// Sometimes if you attach too quickly after spawning screen it will say the
+		// session does not exist so run a command against it until it works.
+		go func() {
+			ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+			defer cancel()
+			defer close(sess.ready)
+			for {
+				select {
+				case <-sess.ctx.Done():
+					sess.ready <- xerrors.Errorf("session %s is gone", id)
+					return
+				case <-ctx.Done():
+					sess.ready <- xerrors.Errorf("timed out waiting for session %s", id)
+					return
+				default:
+					process, err := execer.Start(ctx, Command{
+						Command: "screen",
+						Args:    []string{"-S", sess.screenID, "-X", "version"},
+						UID:     command.UID,
+						GID:     command.GID,
+						Env:     append(command.Env, "SCREENDIR="+screendir),
+					})
+					if err != nil {
+						sess.ready <- xerrors.Errorf("error waiting for session %s: %w", id, err)
+						return
+					}
+					err = process.Wait()
+					// TODO: Send error if it is anything but "no screen session found".
+					if err == nil {
+						return
+					}
+					time.Sleep(250 * time.Millisecond)
+				}
+			}
+		}()
+	}
+
+	// Block until the server is ready.
+	select {
+	case err := <-sess.ready:
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Refresh the session timeout now.
+	sess.timeout.Reset(options.SessionTimeout)
+
+	// Keep refreshing the session timeout while this connection is alive.
+	heartbeat := time.NewTicker(options.SessionTimeout / 2)
+	go func() {
+		defer heartbeat.Stop()
+		// Reset when the connection closes to ensure the session stays up for the
+		// full timeout.
+		defer sess.timeout.Reset(options.SessionTimeout)
+		for {
+			select {
+			// Stop looping once this request finishes.
+			case <-ctx.Done():
+				return
+			case <-heartbeat.C:
+			}
+			sess.timeout.Reset(options.SessionTimeout)
```

Comment:
it's hard to follow the timeout because it is being set and reset in multiple functions, possibly with different timeout lengths.

how about pulling this into a `func (s *session) heartbeat(ctx context.Context)` and having the timeout length be set just once and stored on the session.